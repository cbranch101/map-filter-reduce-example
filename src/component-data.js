import {
  StyledBaseButton,
  StyledLoadingSpinner,
  StyledLoadingSpinnerContainer,
  StyledStartEnhancer,
  StyledEndEnhancer,
  KIND,
} from "baseui/button";

// important details
// the first option is always the default

export const $kind = {
  prop: "$kind",
  defaultOption: "primary",
  options: Object.values(KIND),
  label: "Kind",
};

export const $shape = {
  prop: "$shape",
  defaultOption: "default",
  options: ["default", "pill", "round", "circle", "square"],
  label: "Shape",
};

export const propMap = {
  $isFocusVisible: [false, true],
  $disabled: [false, true],
  $isSelected: [false, true],
  $kind: ["primary", "secondary", "tertiary", "minimal"],
  $size: ["default", "compact", "mini", "large"],
  $shape: ["default", "pill", "round", "circle", "square"],
};

export const defaultProps = Object.keys(propMap).reduce((memo, propName) => {
  return {
    ...memo,
    [propName]: propMap[propName][0],
  };
}, {});

export const componentMap = [
  {
    name: "base-button",
    component: StyledBaseButton,
    concepts: [
      {
        name: "Color States",
        relevantProps: ["$kind", "$isSelected"],
      },
      {
        name: "Container Shape / Size",
        relevantProps: ["$size", "$shape"],
      },
      {
        name: "Focus",
        relevantProps: ["$isFocusVisible"],
      },
    ],
  },
  {
    name: "loading-spinner",
    component: StyledLoadingSpinner,
    concepts: [
      {
        name: "Color States",
        relevantProps: ["$kind", "$disabled", "$isSelected"],
      },
    ],
  },
  {
    name: "loading-spinner-container",
    component: StyledLoadingSpinnerContainer,
    concepts: [
      {
        name: "Color States",
        relevantProps: ["$kind", "$disabled", "$isSelected"],
      },
    ],
  },
  {
    name: "start-enhancer",
    component: StyledStartEnhancer,
  },
  {
    name: "end-enhancer",
    component: StyledEndEnhancer,
  },
];

export default componentMap;
