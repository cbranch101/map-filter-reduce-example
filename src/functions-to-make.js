import { propMap, componentMap } from "./component-data";
import {
  getStylesFromComponent,
  getDifferencesBetweenStyles,
  convertStringToValue,
  getCSSWithClassName,
  convertValueToString,
} from "./style-helpers";

const getKeyFromPropPossibilites = (options) => options.join("_");

const addVariationToPath = (path, propName, propValue) => {
  const separator = path === undefined ? "" : "__";
  const basePath = path === undefined ? "" : path;
  return `${basePath}${separator}${propName}:${convertValueToString(
    propValue
  )}`;
};

const getPropsFromPath = (path) => {
  return path.split("__").reduce((memo, subPath) => {
    const [prop, baseValue] = subPath.split(":");
    const value = convertStringToValue(baseValue);
    return {
      ...memo,
      [prop]: value,
    };
  }, {});
};

const getPathFromProps = (props) => {
  return Object.keys(props).reduce((memo, propName) => {
    const value = props[propName];
    return addVariationToPath(memo, propName, value);
  }, undefined);
};

export const getAllPropVariations = (componentName) => {
  const variations = {};

  const addPropVariationsForComponent = (
    parentName,
    childName,
    propPossibilites,
    Component
  ) => {
    if (!variations[parentName]) {
      variations[parentName] = {};
    }
    if (!variations[parentName][childName]) {
      variations[parentName][childName] = {};
    }

    const combinationKey = getKeyFromPropPossibilites(propPossibilites);

    if (variations[parentName][childName][combinationKey]) {
      return;
    }
    variations[parentName][childName][combinationKey] = {};

    const traversePermutations = (propOptions, path) => {
      const [propName, ...remainingPropOptions] = propOptions;

      const values = propMap[propName];

      values.forEach((value) => {
        const nextPath = addVariationToPath(path, propName, value);
        if (remainingPropOptions.length === 0) {
          const props = getPropsFromPath(nextPath);
          const styles = getStylesFromComponent(Component, props);
          variations[parentName][childName][combinationKey][nextPath] = styles;
        } else {
          traversePermutations(remainingPropOptions, nextPath);
        }
      });
    };
    traversePermutations(propPossibilites);
  };

  componentMap.forEach(
    ({ name: subComponentName, concepts, component: Component }) => {
      if (concepts) {
        concepts.forEach(({ relevantProps }) => {
          addPropVariationsForComponent(
            componentName,
            subComponentName,
            relevantProps,
            Component
          );
        });
      }
    }
  );
  return variations;
};

export const allStyleVariations = getAllPropVariations("Button");

export const getStyleComparisons = (
  possibilites,
  variations,
  defaultProps,
  comparisonStyles,
  comparisonProps = {},
  depth = 1
) => {
  const [propName, ...remainingPossibilities] = possibilites;
  const values = propMap[propName];
  return values.reduce((memo, value, index) => {
    const isTopLevelDefault = depth === 1 && index === 0;
    const nextProps = {
      ...comparisonProps,
      [propName]: value,
    };
    const filledInProps = {
      ...defaultProps,
      ...nextProps,
    };
    const nextStyles = getStylesFromVariations(filledInProps, variations);

    const comparison = {
      props: filledInProps,
      diff: isTopLevelDefault
        ? getDifferencesBetweenStyles({}, comparisonStyles)
        : getDifferencesBetweenStyles(nextStyles, comparisonStyles),
      isDefault: isTopLevelDefault,
    };

    if (remainingPossibilities.length === 0) {
      return [...memo, comparison];
    }

    return [
      ...memo,
      comparison,
      ...getStyleComparisons(
        remainingPossibilities,
        variations,
        defaultProps,
        nextStyles,
        nextProps,
        depth + 1
      ),
    ];
  }, []);
};

const getStylesFromVariations = (props, variations) => {
  const path = getPathFromProps(props);
  return variations[path];
};

const getDefaultProps = (propOptions) => {
  return propOptions.reduce((memo, propName) => {
    return {
      ...memo,
      [propName]: propMap[propName][0],
    };
  }, {});
};

export const getStyleComparisonsForConcept = (
  componentName,
  subComponentName,
  concept
) => {
  const { relevantProps } = concept;
  const conceptKey = getKeyFromPropPossibilites(relevantProps);

  const variations =
    allStyleVariations[componentName][subComponentName][conceptKey];

  const defaultProps = getDefaultProps(relevantProps);

  const defaultStyles = getStylesFromVariations(defaultProps, variations);

  return getStyleComparisons(
    relevantProps,
    variations,
    defaultProps,
    defaultStyles
  );
};

const serializeComparison = (componentName, comparison, defaultProps) => {
  const { diff, props } = comparison;
  if (Object.keys(diff).length === 0) {
    return null;
  }
  return getCSSWithClassName(componentName, props, diff, defaultProps);
};

export const serializeComparisons = (
  componentName,
  comparisons,
  defaultProps
) => {
  const comparisonStrings = comparisons
    .map((comparison) => {
      return serializeComparison(componentName, comparison, defaultProps);
    })
    .filter((string) => !!string);
  return comparisonStrings.join("\n\n");
};

export const getStyleSheetForConcept = (
  componentName,
  subComponentName,
  concept
) => {
  const comparisons = getStyleComparisonsForConcept(
    componentName,
    subComponentName,
    concept
  );
  const { relevantProps } = concept;
  const defaultProps = getDefaultProps(relevantProps);
  return serializeComparisons(subComponentName, comparisons, defaultProps);
};
