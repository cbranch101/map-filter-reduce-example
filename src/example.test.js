import { StyledBaseButton } from "baseui/button";
import {
  getStylesFromComponent,
  getCssFromStyles,
  getCSSFromComponent,
} from "./style-helpers";
import { getStyleSheetForConcept } from "./functions-to-make";

import { defaultProps, componentMap } from "./component-data";
const defaultStyles = getStylesFromComponent(StyledBaseButton, defaultProps);

describe("style-helpers", () => {
  describe("getStylesFromComponent", () => {
    test("should return a style object given a component and props", () => {
      const styles = getStylesFromComponent(StyledBaseButton, defaultProps);
      expect(styles).toMatchSnapshot();
    });
  });
  describe("getCssFromStyles", () => {
    test("should return a valid css style string for the provided styles", () => {
      const css = getCssFromStyles(defaultStyles);
      expect(css).toMatchSnapshot();
    });
  });
  describe("getCSSFromComponent", () => {
    test("should return valid css given a component and props", () => {
      const css = getCSSFromComponent(StyledBaseButton, defaultProps);
      expect(css).toMatchSnapshot();
    });
  });
});

describe("Functions to make", () => {
  describe("getStyleSheetForConcept", () => {
    test("should generate a readable stylesheet, based on all prop variations", () => {
      const sheet = getStyleSheetForConcept(
        "Button",
        "base-button",
        componentMap[0].concepts[0]
      );
      expect(sheet).toMatchSnapshot();
    });
  });
});
