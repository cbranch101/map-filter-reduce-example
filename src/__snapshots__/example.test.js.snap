// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Functions to make getStyleSheetForConcept should generate a readable stylesheet, based on all prop variations 1`] = `
".base-button {
    display:inline-flex;
    flex-direction:row;
    align-items:center;
    justify-content:center;
    border-left-width:0;
    border-top-width:0;
    border-right-width:0;
    border-bottom-width:0;
    border-left-style:none;
    border-top-style:none;
    border-right-style:none;
    border-bottom-style:none;
    outline:none;
    box-shadow:none;
    text-decoration:none;
    -webkit-appearance:none;
    transition-property:background;
    transition-duration:200ms;
    transition-timing-function:cubic-bezier(0, 0, 1, 1);
    cursor:pointer;
    :disabled {
        cursor:not-allowed;
        background-color:#F6F6F6;
        color:#AFAFAF;
    }
    margin-left:0;
    margin-top:0;
    margin-right:0;
    margin-bottom:0;
    font-family:system-ui, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif;
    font-size:16px;
    font-weight:500;
    line-height:20px;
    border-top-right-radius:0px;
    border-bottom-right-radius:0px;
    border-top-left-radius:0px;
    border-bottom-left-radius:0px;
    padding-top:14px;
    padding-bottom:14px;
    padding-left:16px;
    padding-right:16px;
    color:#FFFFFF;
    background-color:#000000;
    :hover {
        background-color:#333333;
    }
    :active {
        background-color:#545454;
    }
}

.base-button.is-selected {
    background-color:#000000;
    :hover {
        background-color:#333333;
    }
    :active {
        background-color:#545454;
    }
}

.base-button.secondary {
    color:#FFFFFF;
    background-color:#000000;
    :hover {
        background-color:#333333;
    }
    :active {
        background-color:#545454;
    }
}

.base-button.secondary.is-selected {
    color:#000000;
    background-color:#EEEEEE;
    :hover {
        background-color:#E2E2E2;
    }
    :active {
        background-color:#CBCBCB;
    }
}

.base-button.tertiary {
    color:#FFFFFF;
    background-color:#000000;
    :hover {
        background-color:#333333;
    }
    :active {
        background-color:#545454;
    }
}

.base-button.tertiary.is-selected {
    background-color:transparent;
    :hover {
        background-color:#F6F6F6;
    }
    :active {
        background-color:#EEEEEE;
    }
}

.base-button.minimal {
    color:#FFFFFF;
    background-color:#000000;
    :hover {
        background-color:#333333;
    }
    :active {
        background-color:#545454;
    }
}

.base-button.minimal.is-selected {
    background-color:transparent;
    :hover {
        background-color:#F6F6F6;
    }
    :active {
        background-color:#EEEEEE;
    }
}"
`;

exports[`style-helpers getCSSFromComponent should return valid css given a component and props 1`] = `
"display:inline-flex;
flex-direction:row;
align-items:center;
justify-content:center;
border-left-width:0;
border-top-width:0;
border-right-width:0;
border-bottom-width:0;
border-left-style:none;
border-top-style:none;
border-right-style:none;
border-bottom-style:none;
outline:none;
box-shadow:none;
text-decoration:none;
-webkit-appearance:none;
transition-property:background;
transition-duration:200ms;
transition-timing-function:cubic-bezier(0, 0, 1, 1);
cursor:pointer;
:disabled {
    cursor:not-allowed;
    background-color:#F6F6F6;
    color:#AFAFAF;
}
margin-left:0;
margin-top:0;
margin-right:0;
margin-bottom:0;
font-family:system-ui, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif;
font-size:16px;
font-weight:500;
line-height:20px;
border-top-right-radius:0px;
border-bottom-right-radius:0px;
border-top-left-radius:0px;
border-bottom-left-radius:0px;
padding-top:14px;
padding-bottom:14px;
padding-left:16px;
padding-right:16px;
color:#FFFFFF;
background-color:#000000;
:hover {
    background-color:#333333;
}
:active {
    background-color:#545454;
}"
`;

exports[`style-helpers getCssFromStyles should return a valid css style string for the provided styles 1`] = `
"display:inline-flex;
flex-direction:row;
align-items:center;
justify-content:center;
border-left-width:0;
border-top-width:0;
border-right-width:0;
border-bottom-width:0;
border-left-style:none;
border-top-style:none;
border-right-style:none;
border-bottom-style:none;
outline:none;
box-shadow:none;
text-decoration:none;
-webkit-appearance:none;
transition-property:background;
transition-duration:200ms;
transition-timing-function:cubic-bezier(0, 0, 1, 1);
cursor:pointer;
:disabled {
    cursor:not-allowed;
    background-color:#F6F6F6;
    color:#AFAFAF;
}
margin-left:0;
margin-top:0;
margin-right:0;
margin-bottom:0;
font-family:system-ui, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif;
font-size:16px;
font-weight:500;
line-height:20px;
border-top-right-radius:0px;
border-bottom-right-radius:0px;
border-top-left-radius:0px;
border-bottom-left-radius:0px;
padding-top:14px;
padding-bottom:14px;
padding-left:16px;
padding-right:16px;
color:#FFFFFF;
background-color:#000000;
:hover {
    background-color:#333333;
}
:active {
    background-color:#545454;
}"
`;

exports[`style-helpers getStylesFromComponent should return a style object given a component and props 1`] = `
Object {
  ":active": Object {
    "backgroundColor": "#545454",
  },
  ":disabled": Object {
    "backgroundColor": "#F6F6F6",
    "color": "#AFAFAF",
    "cursor": "not-allowed",
  },
  ":hover": Object {
    "backgroundColor": "#333333",
  },
  "WebkitAppearance": "none",
  "alignItems": "center",
  "backgroundColor": "#000000",
  "borderBottomLeftRadius": "0px",
  "borderBottomRightRadius": "0px",
  "borderBottomStyle": "none",
  "borderBottomWidth": 0,
  "borderLeftStyle": "none",
  "borderLeftWidth": 0,
  "borderRightStyle": "none",
  "borderRightWidth": 0,
  "borderTopLeftRadius": "0px",
  "borderTopRightRadius": "0px",
  "borderTopStyle": "none",
  "borderTopWidth": 0,
  "boxShadow": "none",
  "color": "#FFFFFF",
  "cursor": "pointer",
  "display": "inline-flex",
  "flexDirection": "row",
  "fontFamily": "system-ui, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
  "fontSize": "16px",
  "fontWeight": 500,
  "justifyContent": "center",
  "lineHeight": "20px",
  "marginBottom": 0,
  "marginLeft": 0,
  "marginRight": 0,
  "marginTop": 0,
  "outline": "none",
  "paddingBottom": "14px",
  "paddingLeft": "16px",
  "paddingRight": "16px",
  "paddingTop": "14px",
  "textDecoration": "none",
  "transitionDuration": "200ms",
  "transitionProperty": "background",
  "transitionTimingFunction": "cubic-bezier(0, 0, 1, 1)",
}
`;
